### ISSUE TITLE
docker commit slow on btrfs storage driver

### ISSUE THREAD

**Issue by umatomba** [docker/docker#10091]
**Description of problem:**
I am using btrfs as a storage driver, because my containers can contain up to 60GB of new data. I thought btrfs, being a copy on write file system, should be able to handle 60GB easy and fast.
In reality docker needs up to 40 minutes to commit a container with 60GB of new data.

In log we see that committing a container with btrfs storage is actually doing some 'Start untar layer' process:

```
    [debug] server.go:999 Calling POST /commit
    2015/01/13 09:06:39 POST /v1.12/commit?author=&comment=&container=b7b6d2056c71&repo=day13_log&tag=
    [0ace70fe] +job commit(b7b6d2056c71)
    [debug] image.go:90 Start untar layer
    [debug] archive.go:75 [tar autodetect] n: [111 117 116 0 0 0 0 0 0 0]
    [debug] image.go:94 Untar time: 2400.110357412s
    [0ace70fe] -job commit(b7b6d2056c71) = OK (0)
```

But all modifications during container-lifetime happen in the btrfs subvolume already, hence the commit should be as fast as a btrfs snapshot is.

**How reproducible:**
1. docker run -itd ubuntu:14.04 /bin/bash
2. make big changes on the container
3. docker commit CONTAINER IMAGE_NAME

`docker version`:
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 5bc2ff8

`docker info`:
Containers: 477
Images: 649
Storage Driver: btrfs
Execution Driver: native-0.2
Kernel Version: 3.13.0-43-generic
Operating System: Ubuntu 14.04.1 LTS
CPUs: 8
Total Memory: 31.1 GiB
ID: 
Debug mode (server): true
Debug mode (client): false
Fds: 24
Goroutines: 30
EventsListeners: 0
Init Path: /usr/bin/docker
Docker Root Dir: /docker
WARNING: No swap limit support

`uname -a`:
Linux lx119-tpo 3.13.0-43-generic #72-Ubuntu SMP Mon Dec 8 19:35:06 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

`lsb_release -a`:
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.1 LTS
Release:        14.04
Codename:       trusty

**Comment by LK4D4** [comment]
@umatomba Hello, this probably was fixed in #9720 
Please try master when you can and report results. Thanks!

**Comment by umatomba** [comment]
Thanks @LK4D4. I will try it now

**Comment by umatomba** [comment]
Hi Alexander,

the last version has the same timings and we see again the same tar operations:

1.) docker run -itd ubuntu:14.04 /bin/bash
8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890

2.) go to docker/btrfs/subvolumes:

```
du -hs 8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890
192M    8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890
```

3.) is it btrfs subvolume? yes.

```
btrfs subvolume list ./| grep 8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890
ID 8952 gen 16456 top level 5 path btrfs/subvolumes/8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890-init
ID 8953 gen 16462 top level 5 path btrfs/subvolumes/8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890
```

4.) docker attach 8d68ec2b5603d3c and make 10Gb changes (one file changes -> to make this test quicker).

5.) go to docker/btrfs/subvolumes:

```
du -hs 8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890
11G    8d68ec2b5603d3c1c1b091ef36bf940998e7ed40589d2f4fac493b4b514ea890
```

6.) time docker commit 8d68ec2b5603 somename:

```
real    2m40.484s
user    0m0.014s
sys     0m0.009s
```

7.) and in docker debug log we see tar...:
```
DEBU[0538] Calling POST /commit
INFO[0538] POST /v1.16/commit?author=&comment=&container=8d68ec2b5603&repo=last_test&tag=
INFO[0538] +job commit(8d68ec2b5603)
DEBU[0538] [tar autodetect] n: [118 97 114 47 0 0 0 0 0 0]
DEBU[0538] Start untar layer
DEBU[0538] [tar autodetect] n: [118 97 114 47 0 0 0 0 0 0]
DEBU[0668] Calling GET /info
INFO[0668] GET /v1.16/info
INFO[0668] +job info()
INFO[0668] +job subscribers_count()
INFO[0668] -job subscribers_count() = OK (0)
INFO[0668] +job registry_config()
INFO[0668] -job registry_config() = OK (0)
INFO[0668] -job info() = OK (0)
DEBU[0676] Calling GET /version
INFO[0676] GET /v1.16/version
INFO[0676] +job version()
INFO[0676] -job version() = OK (0)
DEBU[0698] Untar time: 160.135468021s
INFO[0698] -job commit(8d68ec2b5603) = OK (0)
```

`docker version:`
```
Client version: 1.4.1-dev
Client API version: 1.16
Go version (client): go1.4
Git commit (client): c9dab70
OS/Arch (client): linux/amd64
Server version: 1.4.1-dev
Server API version: 1.16
Go version (server): go1.4
Git commit (server): c9dab70
```

**Comment by LK4D4** [comment]
@umatomba Thanks for testing! But I'm not expert. Summoning expert: @unclejack

**Comment by jfrazelle** [comment]
Also I think for 60GB 40 minutes is not bad... but I am no expert either

**Comment by umatomba** [comment]
@jfrazelle using btrfs COW mechanism it could be <1 minute. I think that current docker version is very AUFS oriented. Is it so?
ping @jpetazzo, @unclejack,  @jlhawn, @alexlarsson

**Comment by cpuguy83** [comment]
You're right, afaict the btrfs driver is using the default/non-optimized differ.

**Comment by paralin** [comment]
Has there been any progress towards fixing this issue? I'm experiencing very slow commits on btrfs at the moment.

**Comment by hellonico** [comment]
Around 40 minutes here too .... container is ~40Go. disk is a fast SSD so this seems a bit slow.

**Comment by philipbergen** [comment]
Same deal here, version 1.10.3.

**Comment by asj** [comment]
If I am not wrong the test case here would translate to [1] in btrfs. And it does not seems to reproduce the problem but I am trying with a 3gfile.

[1]
btrfs su snap /docker/image /docker/image-init
btrfs su snap /docker/image-init /docker/container
cp /60GB-file /docker/container
btrfs su snap /docker/container /btrfs/60gb-file-image

Also can you narrow down using strace ?

**Comment by micgrivas** [comment]
@umatomba , not on the issue, however I think you misuse docker. You should NOT save data in the docker. It is very weird/wrong to see a docker 60 GB. Data should go to volumes, outside docker. docker should contain ONLY the libraries and programs, not the data files, even if they are important to the application.

**Comment by dark-penguin** [comment]
Still an issue in 2020.
Is there any other reason why Docker uses btrfs if not specifically for this?

**Comment by jpetazzo** [comment]
When you do `docker commit`, it does more than just taking a snapshot of the container. It also creates a new image from that container, which involves:
1) computing the "diff" between the container's image and the container (what you see when you run `docker diff`)
2) creating a layer with that diff (a layer is basically a tarball containing all these files)
3) compressing that layer

When using AUFS or overlay, steps 1 and 2 are easy to implement, because the diff is essentially a separate directory. "Computing the diff" means listing that directory, and creating the layer is _almost_ as simple as creating a tar out of that directory (with special handling of _tombstones_ or _whiteouts_, i.e. files that were removed). The cost of these steps is proportional to the amount of data in the read-write layer, but the size of the original image doesn't matter. In other words, if you have a 10 GB image and change a couple of small files, `docker commit` (on AUFS or overlay) should be very fast.

When using another storage driver (like BTRFS, ZFS, or device mapper), step 1 requires to find out all the files that changed between the image and the container. The naive implementation is to walk the entire filesystem, and to be correct, it requires to also compare the content of the files (since even if their size and modification time are identical, nothing guarantees that their content is the same). With BTRFS, it is possible to take a snapshot, and then examine the content of the snapshot to find out which files were modified in that snapshot. This reduces the cost of step 1 (computing the diff). I do not know if Docker uses that optimization.

However, step 2 (building a tar with all the changes) still has to be done, and the cost is proportional to the size of the layer. Sure, taking a BTRFS snapshot might take a few seconds (or even less than that), but then Docker still has to create a tar with all that data.

And step 3 will also be quite expensive, because (if I remember correctly) Docker uses the xz compression algorithm to reduce network usage, and xz compression is very expensive.

The bottom line is that using `docker commit` just to take a snapshot of a container is, unfortunately, a bad idea. The goal of `docker commit` is to create a new image. If you want to take a snapshot, the best solution would be to store your data in a volume (this is considered to be the best practice with Docker anyway; this is not specific to BTRFS), and then take a snapshot of that volume.

I believe that there are _volume plugins_ that might even automate and streamline that, but I haven't looked into that for quite a while now so I'll let you investigate that if that's what you're after.

**Comment by dark-penguin** [comment]
I would assume that:

(3) Compressing the layer is not necessary (or rather harmful) on btrfs, because it defeats the purpose of using btrfs in the first place; the archive will often be [much] bigger than the copy-on-write difference, and if you want compression, you use btrfs compression.

(2) "Creating a layer" is as simple as taking a snapshot - and with btrfs, that layer also includes all the previous layers, so you don't have to worry about diffs. I thought this was the reason for using btrfs for managing layers! Btrfs was literally made for this!

(1) When you commit a (usually already stopped) container, usually before removing it, you probably will not want to see its diff anymore, and you don't need that diff for layer management because btrfs takes care of that for you.

I understand that the developers' thinking was probably different, and most likely for a good reason, but I can't imagine a good reason to tar btrfs subvolumes instead of relying on snapshots to do all the work for you (usually in less than a second). That's the whole point of btrfs.

Searching for any information on this topic yielded this: https://docs.docker.com/storage/storagedriver/btrfs-driver/#how-the-btrfs-storage-driver-works . According to this, it is supposed to work the way I imagined: this page explains how everything is stored as subvolumes and snapshots, and is very efficient. But according to you (and our experience with this issue), it is not. And I couldn't find the place in the sources where this is happening...

Then I've found #443 (apparently, you were actually the btrfs backend implementation author?). Why insist on file-based layers when it is basically reimplementing what btrfs already does wery efficiently, but in a very inefficient way? Not only that, but apparently they are making it even worse, duplicating all the CoW files by unpacking tars instead of taking snapshots...

**Comment by dark-penguin** [comment]
Also this: https://github.com/moby/moby/issues/10161
I thought the switch to btrfs happened because aufs is a slow crutch for what btrfs does natively and instantly. Here we see people talking about using `btrfs send` to get the diffs, and that btrfs is actually many times slower at what it was designed for than that crutch...

**Comment by darkdragon-001** [comment]
I guess that the steps described by @jpetazzo are needed for image creation since images have a common format used to upload/push it for example to hub.docker.com. If this is the case, one could discuss it would be better to store images in a storage driver specific format locally and only convert to the common format when needed (e.g. upload).

**Comment by dark-penguin** [comment]
Naturally. I also assume that it's probably for compatibility: maybe you can switch between storage drivers without having to rebuild all your images? This would mean dumbing down btrfs so that it would mimick aufs, only for this extremely rare and mostly pointless use case. And even in this case, switching from btrfs to aufs would mean you have to try `btrfs sub del` an image if `rmdir` did not work; but I can't imagine switching the other way would work, since it would have to turn existing directories into subvolumes.

An obvious alternative solution would be using `cp -a --reflink=always` instead of subvolumes or tar; this would eliminate all of those problems at the price of no compression. By the way, I don't think they are stored in a compressed state, because `docker pull` clearly shows "Pulling... Extracting...". So they are actually stored extracted. Meaning, there is absolutely nothing preventing either snapshots or copying with reflinks, and probably even switching between storage drivers.

**Comment by jpetazzo** [comment]
I didn't implement the BTRFS storage driver; but I helped. I was part of the "original" Docker team but I haven't worked on the Docker core since 2015 (I think?) and have left Docker Inc in 2018. So my information might be outdated :-)

I should have emphasized in my previous reply: the goal of `docker commit` is not to create a snapshot of a running container. It is primarily to create a new image. As a bonus, yes, it can be used to create a snapshot; but it's more a side effect. That's why it's not always the most efficient (it has to execute steps 2 and 3 in my list above, and these steps are expensive). About the image format, it is used both locally and on registries; so that's why even if you use BTRFS, you still have to build these tarballs, because that's required to push/pull images.

In the early days of Docker, BTRFS was an interesting alternative to AUFS, because AUFS wasn't widely supported. Nowadays the recommended backend is overlay.

As I was saying above - if you want to use the power of BTRFS snapshots, it is probably better to do the snapshots outside of Docker (i.e. not using `docker commit`), using e.g. a volume. Just my 2c!

**Comment by darkdragon-001** [comment]
How are downloaded images stored in the Btrfs storage driver? Are they stored as the compressed diffs and when a container is created a new btrfs subvolume is created for every layer? Or are they always stored as both, compressed diff _and_ btrfs subvolume (taking up twice the space)?

**Comment by jpetazzo** [comment]
Images always get stored twice, if I remember correctly. The compressed diff (to be available in case we want to `docker push` that image in the future) and the uncompressed version (in a format that depends on the specific storage driver).

`docker build` essentially does the equivalent of a `docker commit` after each step, so yes, it does the same operations.

If you want the snapshotting ability without the overhead, _perhaps_ you can have a look at containerd and its "snapshotters". Full disclosure: I have limited first-hand experience with containerd, so I have actually no idea if it would let you do that; but if I were looking for something like this (while retaining the convenience of an API, CLI, etc) I'd start looking there maybe!

**Comment by darkdragon-001** [comment]
Thanks for your insights.

I guess that most setups _either_ consume an image locally _or_ want to publish it (via push). So I think it makes sense to decouple those two (this would speed up build times significantly and reduce cost of CI machines significantly)!

**Comment by dark-penguin** [comment]
I understand that you're not supposed to keep your changes inside the container and commit it often, but yes, during `docker build`, starting and stopping containers takes more time than actually building them. Which is why people are forced to merge multiple commands into one long multi-line command, which defeats the reason for committing the intermediate containers in the first place...

It would certainly make more sense to generate a tar when you do `docker push`. That happens much less often than committing, and it's not a big deal if it takes a little more time - but it is a big problem if we have to slow most other operations and take additional space just to be always ready to marginally improve the speed of `docker push`.

I took a look at alternative solutions. According to my knowledge, Podman is known to be "Docker that's less stubborn" - they "fix" things Docker refuses to fix allegedly for ideological reasons (like init system support). But no, Podman does the same; even worse, it builds the commit tar in /tmp, which I think is a terrible idea (it's often not large enough, and this means thrashing not only the Docker drive, but also the tmp drive).

One solution that to my surprise I really like so far is systemd-container. I don't think it adheres to any standard - systemd is not known for that - but I think it is much easier to manage the way you want. It supports an init system (though it really prefers that init system to be systemd), it is meant to be used for services that are autostarted and managed by systemd, and it can run from any folder you want. For example, mount tmpfs somewhere, extract your whole image there, use it at a blazing speed, and then write it back to /var/lib/ or wherever you want. Even with systemd inside of it, it starts and stops several times faster than Docker, and it actually uses btrfs *snapshots* by default. I think I'll write my own "snapshotter" script and a set of tools for it, and just use that (what I need is primarily not CI).

**Comment by cpuguy83** [comment]
If you need more flexibility, the. definitely just use containerd.
It gives you full access to the storage layer.

**Comment by thaJeztah** [comment]
@dark-penguin asp worth trying `docker build` with BuildKit enabled (`DOCKER_BUILDKIT=1 docker build .....`), which does not create a new container for every build-step.

**Comment by darkdragon-001** [comment]
@dark-penguin If you are looking for built-in volume snapshots (in contrast to container snapshots), the discussion is happening in #33782.

**Comment by dark-penguin** [comment]
No, I don't want volume snapshots, I just want #41109 . :) I never use "push" at all, so this "optimization" is completely wasted on me. I want no space+performance overhead for tars - only btrfs snapshots.
